File: scripts/gpureg.pl
use strict;
my $GPU_BASE = 0x2000240;
#SW regs
my $GPU_CTRL_REG          = $GPU_BASE + 0x0;
my $GPU_IMEM_ADDR_REG     = $GPU_BASE + 0x4;
my $GPU_IMEM_WDATA_REG    = $GPU_BASE + 0x8;
my $GPU_DMEM_ADDR_REG     = $GPU_BASE + 0xc;
my $GPU_DMEM_WDATA_LO_REG = $GPU_BASE + 0x10;
my $GPU_DMEM_WDATA_HI_REG = $GPU_BASE + 0x14;
my $GPU_PARAM_ADDR_REG    = $GPU_BASE + 0x18;
my $GPU_PARAM_DATA_LO_REG = $GPU_BASE + 0x1c;
my $GPU_PARAM_DATA_HI_REG = $GPU_BASE + 0x20;

#HW dbg regs
my $GPU_PC_DBG_REG        = $GPU_BASE + 0x24;
my $GPU_IF_INSTR_REG      = $GPU_BASE + 0x28;
my $GPU_DMEM_RDATA_LO_REG = $GPU_BASE + 0x2c;
my $GPU_DMEM_RDATA_HI_REG = $GPU_BASE + 0x30;
my $GPU_DONE              = $GPU_BASE + 0x34;

##########################################################################
####################### READ/WRITE HELPERS ###############################
###########################################################################
sub regwrite {
  my($addr, $value) = @_;
  my $cmd = sprintf("regwrite 0x%08x 0x%08x", $addr, $value);
  `$cmd`;
}

sub regread {
  my($addr) = @_;
  my $cmd = sprintf("regread 0x%08x", $addr);
  my @out = `$cmd`;
  my $result = $out[0];
  if ($result =~ m/Reg (0x[0-9a-f]+) \((\d+)\):\s+(0x[0-9a-f]+) \((\d+)\)/i) {
    return $3;
  }
  return $result;
}
# OUTPUT: Reg 0xADDR (DEC): 0xVALUE (DEC)

###########################################################################
############################ GPU REG CMDS ################################
###########################################################################
sub usage {
  print "Usage: gpureg <cmd> [args]\n";
  print "  Commands:\n";
  print "    run <0|1>                                   set run\n";
  print "    step                                        single step\n";
  print "    pcreset                                     pc_reset_pulse\n";
  print "    imem_write <addr> <wdata>                   program I-mem word\n";
  print "    dmem_write <addr> <hi> <lo>                 program D-mem 64b\n";
  print "    dmem_read <addr>                            read D-mem 64b via portB\n";
  print "    dbg                                         print pc + if_instr\n";
  print "    allregs                                     dump all hw regs\n";
  print "    param_write <addr> <hi> <lo>                program param_write 64b\n";
}


  # wire run_level      =  sw_ctrl[0];
  # wire step           =  sw_ctrl[1];
  # wire pc_reset       =  sw_ctrl[2];
  # wire imem_prog_we   =  sw_ctrl[3];
  # wire dmem_prog_en   =  sw_ctrl[4];
  # wire dmem_prog_we   =  sw_ctrl[5];
  # wire param_wr_en    =  sw_ctrl[6];

sub ctrl_read_val {
  my $v = regread($GPU_CTRL_REG);
  $v =~ s/\s+//g;
  return hex($v);
}

sub ctrl_write_val {
  my($v) = @_;
  regwrite($GPU_CTRL_REG, $v);
}

sub ctrl_set_bit {
  my($bit, $val) = @_;
  my $v = ctrl_read_val();
  if ($val) {
    $v |= (1 << $bit); 
    }
    else{
    $v &= ~(1 << $bit); 
    }
  ctrl_write_val($v);
}

sub ctrl_pulse_bit {
  my($bit) = @_;
  ctrl_set_bit($bit, 0);
  ctrl_set_bit($bit, 1);
  ctrl_set_bit($bit, 0);
}

sub cmd_run {
  my($on) = @_;
  ctrl_set_bit(0, $on ? 1 : 0); # run_level
}

sub cmd_step {
  ctrl_pulse_bit(1); # step_pulse
}

sub cmd_pcreset {
  ctrl_pulse_bit(2); # pc_reset_pulse
}

sub cmd_imem_write {
  my($addr, $wdata) = @_;
  my $a = ($addr =~ /^0x/i) ? hex($addr) : int($addr);
  my $d = ($wdata =~ /^0x/i) ? hex($wdata) : hex("0x$wdata");
  regwrite($GPU_IMEM_ADDR_REG, $a);
  regwrite($GPU_IMEM_WDATA_REG, $d);
  ctrl_pulse_bit(3); # imem_we 
}

sub cmd_dmem_write {
  my($addr, $hi, $lo) = @_;
  my $a  = ($addr =~ /^0x/i) ? hex($addr) : int($addr);
  my $hi_v = ($hi =~ /^0x/i) ? hex($hi) : hex("0x$hi");
  my $lo_v = ($lo =~ /^0x/i) ? hex($lo) : hex("0x$lo");
  regwrite($GPU_DMEM_ADDR_REG, $a);
  regwrite($GPU_DMEM_WDATA_HI_REG, $hi_v);
  regwrite($GPU_DMEM_WDATA_LO_REG, $lo_v);
  ctrl_set_bit(4, 1); # dmem_en=1
  ctrl_set_bit(5, 1); # dmem_we=1
  ctrl_set_bit(5, 0); # dmem_we=0
}

sub cmd_dmem_read {
  my($addr) = @_;
  my $a  = ($addr =~ /^0x/i) ? hex($addr) : int($addr);
  regwrite($GPU_DMEM_ADDR_REG, $a);
  ctrl_set_bit(4, 1); # dmem_en=1
  ctrl_set_bit(5, 0); # dmem_we=0
  my $lo = regread($GPU_DMEM_RDATA_LO_REG);
  my $hi = regread($GPU_DMEM_RDATA_HI_REG);
  print "DMEM[$a] = $hi$lo\n";
}

sub cmd_dbg {
  print "PC:       ", regread($GPU_PC_DBG_REG), "\n";
  print "IF_INSTR: ", regread($GPU_IF_INSTR_REG), "\n";
}

sub cmd_allregs {
  cmd_dbg();
  print "DMEM_RLO: ", regread($GPU_DMEM_RDATA_LO_REG), "\n";
  print "DMEM_RHI: ", regread($GPU_DMEM_RDATA_HI_REG), "\n";
}

sub cmd_param_write {
  my($addr, $hi, $lo) = @_;
  my $a  = ($addr =~ /^0x/i) ? hex($addr) : int($addr);
  my $hi_v = ($hi =~ /^0x/i) ? hex($hi) : hex("0x$hi");
  my $lo_v = ($lo =~ /^0x/i) ? hex($lo) : hex("0x$lo");
  regwrite($GPU_PARAM_ADDR_REG, $a);
  regwrite($GPU_PARAM_DATA_HI_REG, $hi_v);
  regwrite($GPU_PARAM_DATA_LO_REG, $lo_v);
  ctrl_pulse_bit(6); # imem_we= 0->1->0
}

#=======================MAIN===========================
my $numargs = $#ARGV + 1;
if ($numargs < 1) {
  usage();
  exit(1);
  }

my $cmd = $ARGV[0];

if ($cmd eq "run") {
  die "run <0|1>\n" if $numargs < 2;
  cmd_run($ARGV[1]);
}
elsif ($cmd eq "step") {
  cmd_step();
}
elsif ($cmd eq "pcreset") {
  cmd_pcreset();
}
elsif ($cmd eq "imem_write") {
  die "imem_write <addr> <wdata>\n" if $numargs < 3;
  cmd_imem_write($ARGV[1], $ARGV[2]);
}
elsif ($cmd eq "dmem_write") {
  die "dmem_write <addr> <hi> <lo>\n" if $numargs < 4;
  cmd_dmem_write($ARGV[1], $ARGV[2], $ARGV[3]);
}
elsif ($cmd eq "dmem_read") {
  die "dmem_read <addr>\n" if $numargs < 2;
  cmd_dmem_read($ARGV[1]);
}
elsif ($cmd eq "dbg") {
  cmd_dbg();
}
elsif ($cmd eq "allregs") {
  cmd_allregs();
}
elsif ($cmd eq "param_write") {
  die "param_write <addr> <hi> <lo>\n" if $numargs < 4;
  cmd_param_write($ARGV[1], $ARGV[2], $ARGV[3]);
}
else {
  print "Unrecognized command $cmd\n";
  usage();
  exit(1);
}

File: scripts/python/gpureg.py
#!/usr/bin/env python
import sys
import re
import subprocess

GPU_BASE = 0x2000000

# SW regs
GPU_CTRL_REG          = GPU_BASE + 0x0
GPU_IMEM_ADDR_REG     = GPU_BASE + 0x4
GPU_IMEM_WDATA_REG    = GPU_BASE + 0x8
GPU_DMEM_ADDR_REG     = GPU_BASE + 0xc
GPU_DMEM_WDATA_LO_REG = GPU_BASE + 0x10
GPU_DMEM_WDATA_HI_REG = GPU_BASE + 0x14
GPU_PARAM_ADDR_REG    = GPU_BASE + 0x18
GPU_PARAM_DATA_LO_REG = GPU_BASE + 0x1c
GPU_PARAM_DATA_HI_REG = GPU_BASE + 0x20

# HW dbg regs
GPU_PC_DBG_REG        = GPU_BASE + 0x24
GPU_IF_INSTR_REG      = GPU_BASE + 0x28
GPU_DMEM_RDATA_LO_REG = GPU_BASE + 0x2c
GPU_DMEM_RDATA_HI_REG = GPU_BASE + 0x30
GPU_DONE              = GPU_BASE + 0x34

# ctrl reg bit assignments:
#   [0] run_level
#   [1] step
#   [2] pc_reset
#   [3] imem_prog_we
#   [4] dmem_prog_en
#   [5] dmem_prog_we
#   [6] param_wr_en


def write_line(text):
    sys.stdout.write("%s\n" % text)


def regwrite(addr, value):
    cmd = "regwrite 0x%08x 0x%08x" % (addr, value)
    subprocess.call(cmd, shell=True)


def regread(addr):
    cmd = "regread 0x%08x" % addr
    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
    out = proc.communicate()[0]
    if not isinstance(out, str):
        out = out.decode("utf-8")
    result = out.splitlines()[0] if out else ""
    m = re.match(r"Reg (0x[0-9a-f]+) \(\d+\):\s+(0x[0-9a-f]+) \(\d+\)", result, re.IGNORECASE)
    if m:
        return m.group(2)
    return result


def ctrl_read_val():
    v = regread(GPU_CTRL_REG).strip()
    return int(v, 16)


def ctrl_write_val(v):
    regwrite(GPU_CTRL_REG, v)


def ctrl_set_bit(bit, val):
    v = ctrl_read_val()
    if val:
        v |= (1 << bit)
    else:
        v &= ~(1 << bit)
    ctrl_write_val(v)


def ctrl_pulse_bit(bit):
    ctrl_set_bit(bit, 0)
    ctrl_set_bit(bit, 1)
    ctrl_set_bit(bit, 0)


def normalize_number(s):
    return str(s).strip().replace("_", "")


def parse_int(s):
    return int(normalize_number(s), 0)


def parse_word(s):
    token = normalize_number(s)
    if token.startswith("0x") or token.startswith("0X"):
        return int(token, 0)
    return int(token, 16)


def cmd_run(on):
    ctrl_set_bit(0, 1 if on else 0)


def cmd_step():
    ctrl_pulse_bit(1)


def cmd_pcreset():
    ctrl_pulse_bit(2)


def cmd_imem_write(addr, wdata):
    a = parse_int(addr)
    d = parse_word(wdata)
    regwrite(GPU_IMEM_ADDR_REG, a)
    regwrite(GPU_IMEM_WDATA_REG, d)
    ctrl_pulse_bit(3)


def cmd_dmem_write(addr, hi, lo):
    a    = parse_int(addr)
    hi_v = parse_word(hi)
    lo_v = parse_word(lo)
    regwrite(GPU_DMEM_ADDR_REG, a)
    regwrite(GPU_DMEM_WDATA_HI_REG, hi_v)
    regwrite(GPU_DMEM_WDATA_LO_REG, lo_v)
    ctrl_set_bit(4, 1)
    ctrl_set_bit(5, 1)
    ctrl_set_bit(5, 0)


def cmd_dmem_read(addr):
    a = parse_int(addr)
    regwrite(GPU_DMEM_ADDR_REG, a)
    ctrl_set_bit(4, 1)
    ctrl_set_bit(5, 0)
    lo = regread(GPU_DMEM_RDATA_LO_REG)
    hi = regread(GPU_DMEM_RDATA_HI_REG)
    write_line("DMEM[%s] = %s%s" % (a, hi, lo))


def cmd_dbg():
    write_line("PC:       %s" % regread(GPU_PC_DBG_REG))
    write_line("IF_INSTR: %s" % regread(GPU_IF_INSTR_REG))


def cmd_allregs():
    cmd_dbg()
    write_line("DMEM_RLO: %s" % regread(GPU_DMEM_RDATA_LO_REG))
    write_line("DMEM_RHI: %s" % regread(GPU_DMEM_RDATA_HI_REG))


def cmd_param_write(addr, hi, lo):
    a    = parse_int(addr)
    hi_v = parse_word(hi)
    lo_v = parse_word(lo)
    regwrite(GPU_PARAM_ADDR_REG, a)
    regwrite(GPU_PARAM_DATA_HI_REG, hi_v)
    regwrite(GPU_PARAM_DATA_LO_REG, lo_v)
    ctrl_pulse_bit(6)


def cmd_done_check():
    write_line("DONE: %s" % regread(GPU_DONE))


def usage():
    write_line("Usage: gpureg.py <cmd> [args]")
    write_line("  Commands:")
    write_line("    run <0|1>                                   set run")
    write_line("    step                                        single step")
    write_line("    pcreset                                     pc_reset_pulse")
    write_line("    imem_write <addr> <wdata>                   program I-mem word")
    write_line("    dmem_write <addr> <hi> <lo>                 program D-mem 64b")
    write_line("    dmem_read <addr>                            read D-mem 64b via portB")
    write_line("    dbg                                         print pc + if_instr")
    write_line("    allregs                                     dump all hw regs")
    write_line("    param_write <addr> <hi> <lo>                program param_write 64b")
    write_line("    done_check                                  check done register")


def run_command(args):
    if not args:
        usage()
        return 1

    cmd = args[0]

    if cmd == "run":
        if len(args) < 2:
            raise SystemExit("run <0|1>")
        cmd_run(int(args[1]))
    elif cmd == "step":
        cmd_step()
    elif cmd == "pcreset":
        cmd_pcreset()
    elif cmd == "imem_write":
        if len(args) < 3:
            raise SystemExit("imem_write <addr> <wdata>")
        cmd_imem_write(args[1], args[2])
    elif cmd == "dmem_write":
        if len(args) < 4:
            raise SystemExit("dmem_write <addr> <hi> <lo>")
        cmd_dmem_write(args[1], args[2], args[3])
    elif cmd == "dmem_read":
        if len(args) < 2:
            raise SystemExit("dmem_read <addr>")
        cmd_dmem_read(args[1])
    elif cmd == "dbg":
        cmd_dbg()
    elif cmd == "allregs":
        cmd_allregs()
    elif cmd == "param_write":
        if len(args) < 4:
            raise SystemExit("param_write <addr> <hi> <lo>")
        cmd_param_write(args[1], args[2], args[3])
    elif cmd == "done_check":
        cmd_done_check()
    else:
        write_line("Unrecognized command: %s" % cmd)
        usage()
        return 1
    return 0


def main():
    sys.exit(run_command(sys.argv[1:]))


if __name__ == "__main__":
    main()
